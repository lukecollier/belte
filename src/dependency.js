import { keys, zipWith, map, values, flatten, concat, prepend, fromPairs } from 'ramda';
import pathUtil from 'path';
import jsx from 'acorn-jsx';
import * as acorn from 'acorn';
import fs from 'fs';
import util from 'util';
import { walk } from 'estree-walker';
import nodeResolve from 'resolve';

/**
 * Resolves an component file using the relevent loader cyclically will generate
 * other components.
 * @param {string} src - The source path for the component.
 * @param {Function} filter - Function explaining how to tell to include in the list or not 
 * component.
 * @returns {(string|Array)} - A list of dependent client scripts in call order.
 */
export const resolve = (src, filter, transform = (_) => _) => {
	const sources = imports(src, filter, transform);
	const recurse = map(({ src }) => resolveAsList(src, filter, transform), sources);
  return prepend(src, map(({src})=>src, flatten(concat(sources, recurse))));
}

/**
 * Resolves an component file using the relevent loader cyclically will generate
 * other components.
 * @param {string} src - The source path for the component.
 * @param {Function} filter - Function explaining how to tell to include in the list or not 
 * component.
 * @returns {(string|Array)} - A list of dependent client scripts in call order.
 */
export const resolveAsList = (src, filter, transform = (_) => _) => {
	const sources = imports(src, filter, transform);
	const recurse = map(({ src }) => resolveAsList(src, filter, transform), sources);
  return flatten(concat(sources, recurse));
}

/**
 * Finds all imports alises and sources for an file 
 * @param {string} src - The source path for the file.
 * @param {Function} filter - Filtering function.
 * @returns {(string|Array)} - A list of dependent client scripts in call order.
 */
export const imports = (src, filter, transform = (_) => _) => {
  const content = transform(fs.readFileSync(src,'utf8'));
  const result = acorn.Parser.extend(jsx()).parse(content, {sourceType: "module"});
  var sources = [];
  walk(result, { enter: function (node, parent ) {
    if (node.type === 'ImportDeclaration') {
      const alias = node.source.value;
      const value = nodeResolve.sync(alias, { basedir: pathUtil.dirname(src) });
      if (filter(value)) {
        sources.push({ src: value, alias: alias }); 
      }
    }
  }});
  return sources;
}
