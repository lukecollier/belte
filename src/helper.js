export const stripExtension = (filename) => {
  return filename.substring(0, filename.indexOf(".")); 
};

export const filenameFromPath = (path) => {
  return path.substring(path.lastIndexOf('/')+1, path.length);
};

export const nameFromPath = (path) => {
  const filename = filenameFromPath(path);
  return stripExtension(filename);
};

export const hyphenCaseToTitleCase = (str) => {
  const lowStr = str.toLowerCase();
  var result = [];
  var lastPos = -1;
  for(var currPos=0; currPos < lowStr.length; currPos++) {
    if (str[currPos] === '-') {
      result.push(lowStr.substring(lastPos + 1, currPos));
      lastPos = currPos;
    }
  }
  result.push(lowStr.substring(lastPos+1, lowStr.length));
  return result.map((lowStr)=>lowStr.charAt(0).toUpperCase() + lowStr.slice(1)).join("");
};
