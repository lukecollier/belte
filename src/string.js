import { head, reduce, tail, toUpper, toLower } from 'ramda';

export const stripExtension = (filename) => {
  const pos = filename.lastIndexOf(".");
  return (pos === -1) ? filename : filename.substring(0, pos);
};

export const filenameFromPath = (path) => {
  return path.substring(path.lastIndexOf('/')+ 1, path.length);
};

export const nameFromPath = (path) => {
  const filename = filenameFromPath(path);
  return stripExtension(filename);
};


export const hypenCaseFromPath = (path) => {
  const isUpper = (ch) => toUpper(ch) === ch;
  const titleCase = nameFromPath(path);
  const lowerCasedHead = toLower(head(titleCase)) + tail(titleCase);
  return reduce((acc, ch) => 
    (isUpper(ch)) ?  `${acc}-${toLower(ch)}` : `${acc}${ch}`, '', lowerCasedHead);
};
