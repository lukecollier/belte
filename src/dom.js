import RewritingStream from 'parse5-html-rewriting-stream';
import parse5 from 'parse5';
import * as dom5 from 'dom5';
import { hyphenCaseToTitleCase } from './string.js';

const HTML5_ELEMENT_NAMES = ['html', 'div', 'base', 'head', 'link', 'meta', 'style', 'title', 'body', 'address', 'article', 'aside', 'footer', 'header', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hgroup', 'main', 'nav', 'section', 'blockquote', 'dd', 'dir', 'dl', 'dt', 'figcaption', 'figure', 'hr', 'li', 'main', 'ol', 'p', 'pre', 'ul', 'a', 'abbr', 'b', 'bdi', 'bdo', 'br', 'cite', 'code', 'data', 'dfn', 'em', 'i', 'kbd', 'mark', 'q', 'rb', 'rtc', 'ruby', 's', 'samp', 'small', 'span', 'strong', 'sub', 'sup','time', 'u', 'var', 'wbr', 'area', 'audio', 'img', 'map', 'track', 'video', 'applet', 'embed', 'iframe', 'noembed', 'object', 'param', 'picture', 'source', 'canvas', 'noscript', 'script', 'del', 'ins', 'caption', 'col', 'colgroup', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr', 'button', 'datalist', 'fieldset', 'form', 'input', 'label', 'legend', 'meter', 'optgroup', 'option', 'output', 'progress', 'select', 'textarea', 'details', 'dialog', 'menu', 'menuitem', 'summary', 'content', 'element', 'shadow', 'slot', 'template'];
export const isCustomElement = (tagNode) => {
  return !HTML5_ELEMENT_NAMES.includes(tagNode.tagName);
}

export const parse = (html) => parse5.parse(html, {scriptingEnabled:false});

export const serialize = (dom) => parse5.serialize(dom);

export const appendToHead = (dom, headElements = []) => {
  const headSet = new Set(headElements);
  const head = dom5.query(dom, (el) => el.tagName === 'head');
  headSet.forEach(elementHtml => 
    dom5.append(head, parse5.parseFragment(elementHtml, {scriptingEnabled:false})));
}

export const domRefs = (dom, resolvers = {Default: (attr) => ''}) => {
  const body = dom5.query(dom, (el) => el.tagName === 'body');
  var refs = new Map();
  dom5.queryAll(body, (el) => isCustomElement(el)).forEach((el, i) => {
    const attribs = {};
    const id = `SvelteComponent-${i}`;
    el.attrs.forEach(attr => attribs[attr.name] = attr.value);
    const name = hyphenCaseToTitleCase(el.tagName);
    if (refs.has(name)) {
      refs.set(name, [...refs.get(name), {id: id, attr: attribs}]);
    } else {
      refs.set(name, [{id: id, attr: attribs}]);
    }
    if (name in resolvers) {
      const newEl = parse5.parseFragment(resolvers[name](attribs), {scriptingEnabled:false});
      const first = dom5.query(newEl, (el) => true);
      first.attrs.push({name: 'id', value: id});
      dom5.replace(el, newEl);
    } else {
      const newEl = parse5.parseFragment(resolvers.Default(attribs), {scriptingEnabled:false});
      const first = dom5.query(newEl, (el) => true);
      first.attrs.push({name: 'id', value: id});
      dom5.replace(el, newEl);
    }
  });
  return refs;
}
