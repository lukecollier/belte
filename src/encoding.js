import Hashids from 'hashids';
import { nameFromPath } from './string.js';

export const encodeStr = (str) => {
  const hashids = new Hashids('salt-me', 8, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ');
  const sumCharacters = [...str]
    .map((_, i)=>str.charCodeAt(i))
    .reduce((accumulator, currentValue) => accumulator + currentValue);
  return hashids.encode(sumCharacters);
}

export const encodeFromPath = (path) => {
  const hashids = new Hashids('salt-me', 8, 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789');
  const sumCharacters = [...path]
    .map((_, i)=>path.charCodeAt(i))
    .reduce((accumulator, currentValue) => accumulator + currentValue);
  return 'Svelte' + '$' + hashids.encode(sumCharacters);
}
