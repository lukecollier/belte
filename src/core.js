import cheerio from 'cheerio';

import fs from 'fs';
import path from 'path';

import { loader } from './svelteLoader';
import { hyphenCaseToTitleCase } from './string';
import { encodeStr } from './encoding';

const createInlineHtml = (name, loader, $, attr) => {
  const src = path.resolve(__dirname, `../__tests__/resource/svelte/${name}.html`);
  const { inline } = loader(src, attr, []);
  const el = $(inline)
  return el; 
};

const createEndScript = (name, loader, params) => {
  const src = path.resolve(__dirname, `../__tests__/resource/svelte/${name}.html`);
  const { end } = loader(src, {}, params);
  return end; 
};

const createHead = (name, loader, params) => {
  const src = path.resolve(__dirname, `../__tests__/resource/svelte/${name}.html`);
  const { head } = loader(src, {}, []);
  var result = new Set();
  for (const entry of head.entries()) {
    write(`${entry[0]}.js`, entry[1]); 
    result.add(`<script src="${entry[0]}.js" defer="true"/>`)
  }
  const compInitScript = createEndScript(name,loader,params);
  write(`${encodeStr(compInitScript)}init.js`, compInitScript); 
  const initHead = `<script src="${encodeStr(compInitScript)}init.js" defer="true"/>`;
  return [...result.values(), initHead];
}

// Has to be a better way to do this~
const HTML5_ELEMENT_NAMES = ['div', 'base', 'head', 'link', 'meta', 'style', 'title', 
'body', 'address', 'article', 'aside', 'footer', 'header', 'h1', 'h2', 'h3', 'h4', 
'h5', 'h6', 'hgroup', 'main', 'nav', 'section', 'blockquote', 'dd', 'dir', 'dl', 
'dt', 'figcaption', 'figure', 'hr', 'li', 'main', 'ol', 'p', 'pre', 'ul', 'a', 'abbr', 
'b', 'bdi', 'bdo', 'br', 'cite', 'code', 'data', 'dfn', 'em', 'i', 'kbd', 'mark', 'q', 
'rb', 'rtc', 'ruby', 's', 'samp', 'small', 'span', 'strong', 'sub', 'sup','time', 'u', 
'var', 'wbr', 'area', 'audio', 'img', 'map', 'track', 'video', 'applet', 'embed', 
'iframe', 'noembed', 'object', 'param', 'picture', 'source', 'canvas', 'noscript', 
'script', 'del', 'ins', 'caption', 'col', 'colgroup', 'table', 'tbody', 'td', 'tfoot', 
'th', 'thead', 'tr', 'button', 'datalist', 'fieldset', 'form', 'input', 'label', 
'legend', 'meter', 'optgroup', 'option', 'output', 'progress', 'select', 'textarea', 
'details', 'dialog', 'menu', 'menuitem', 'summary', 'content', 'element', 'shadow', 
'slot', 'template'];
const isCustomElement = (node) => {
  return node.type === 'tag' && !HTML5_ELEMENT_NAMES.includes(node.name);
}

const write = (filename, data) => {
  const dir = `${process.env.PWD}/build`
  if (!fs.existsSync(dir)) fs.mkdirSync(dir);

  fs.writeFile(`${dir}/${filename}`, data, function(err, data) {
    if (err) throw `Error writing ${filename}`;
    return filename;
  });

  return filename;
}

export const compile = (html, loader) => {
  const $ = cheerio.load(html, { 
    normalizeWhitespace: true, 
    recognizeSelfClosing: true 
  });
  var elRefs = new Map();

  $('body').children().filter((i, node) => isCustomElement(node)).each((i, node) => {
    const name = hyphenCaseToTitleCase(node.name);
    const el = createInlineHtml(name, loader, $, node.attribs);
    const id = `belte-${i}`;
    el.attr('id', `${id}`);
    if (elRefs.has(name)) {
      elRefs.set(name, [...elRefs.get(name), {id: id, attr: node.attribs}]);
    } else {
      elRefs.set(name, [{id: id, attr: node.attribs}]);
    }
    $(node).replaceWith(el);
  });

  elRefs.forEach((params, name, map) => {
    createHead(name,loader, params).forEach(script=>$('head').append($(script)));
  });
  
  write('index.html', $.html());

  return $.html();
};

export const defaultCompile = (html) => {
  return compile(html, loader);
};

export default defaultCompile;
