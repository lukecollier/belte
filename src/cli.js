import { compile } from './core.js' 
import path from 'path';
import {readFileSync, accessSync, constants, mkdirSync, writeFile} from 'fs';
import { forEachObjIndexed } from 'ramda';

// our basic loaders for svelte 2 and react
import * as react from './loader/react.js'; 
import * as sveltev2 from './loader/sveltev2.js'; 

const {R_OK, W_OK} = constants;

var argv = require('minimist')(process.argv.slice(2));

const [first, ...components] = argv._;
if (!argv.help && first === 'on' && components.length !== 0) {
  const template = getNeededAttr('template');
  const output = getNeededAttr('output');
  const templateSrc = path.resolve(process.cwd(), template);
  const target = path.resolve(process.cwd(), output);
  if (canAccessPaths([...components, templateSrc, target])) {
    const data = readFileSync(templateSrc, 'utf8');
    try {
      const opts = {
        components: components.map(src => path.resolve(process.cwd(), src)),
        salt: 'default-salt'
      };

      const loader = getOptionalAttr('loader');
      const { js, html } = compileIt(loader,data, opts, target);
    } catch (e) {
      exitWithError(`failed to compile ${e}`);
    }
  }
} else {
	help();
}

async function compileIt (loader, data, opts, target) {
  const compiled = (loader.err) ? 
    compile(data, opts, react) : compile(data, opts, getLoader(loader.name))
  compiled.then(result => {
    forEachObjIndexed((code, key)=>{
      write(`${target}/${key}.js`, code);
    }, result.js);
    const index = getOptionalAttr('index');
    if (index.err) {
      write(`${target}/index.html`, result.html)
    } else {
      const indexSrc = path.resolve(process.cwd(), index.name);
      write(indexSrc, result.html)
    }
  }).catch(err=>exitWithError(err));
}


function getLoader(loader) {
  switch(loader) {
    case 'react':
      return react; 
    case 'sveltev2':
      return sveltev2; 
    default:
      exitWithError(`loader not supported yet`);
  }
}

function canAccessPaths(paths) {
  return paths.reduce((acc, src) => {
    return acc && tryAccess(src)
  });
}

function tryAccess(src) {
  try {
    accessSync(src, W_OK); 
    return true;
  } catch (_err) {
    exitWithError(`can't access output file at "${src}", do you have access and the file exists?`);
  }
}

function help() {
  console.log('todo: implement help');
}

function getOptionalAttr(attr) {
  const shortName = attr.substring(0,1);
  if (present(attr)) {
    return { name: argv[attr] }
  } else if (present(shortName)) {
    return { name: argv[shortName] }
  } else {
    return { err: `${attr} not given`, name: attr} 
  }
}

function getNeededAttr(attr) {
  const shortName = attr.substring(0,1);
  if (present(attr)) {
    return argv[attr]
  } else if (present(shortName)) {
    return argv[shortName]
  } else {
    exitWithError(`missing needed parameter: ${attr} ${shortName}`);
  }
}

function present(name) {
  return Object.keys(argv).includes(name)
}

function exitWithError(message) {
  console.log('[error]', message);
  process.exit(1);
}

function write(target, data) {
  const src = path.resolve(process.cwd(), target)
  mkdirSync(path.dirname(target), { recursive: true });
  writeFile(target, data, (err) => {
    if (err) {
      exitWithError(err);
    } else {
      console.log('successs!', `written to "${src}"`);
    }});
}
