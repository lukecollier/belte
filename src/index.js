import { compile } from './core.js' 

compile('<html><custom-element></custom-element></html>');

export default compile;
