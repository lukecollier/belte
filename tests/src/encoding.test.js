import { encodeForFileName } from '../../src/encoding';
import { readFileSync } from 'fs';
import pathUtil from 'path';

test('hash is consistent for filesnames', () => {
  const path = pathUtil.resolve(__dirname, '../resource/hashme.txt');
  
  const result = encodeForFileName(readFileSync(path), 'salty');
  expect(result).toMatchSnapshot();
});

test('changing contents changes the hash', () => {
  const path = pathUtil.resolve(__dirname, '../resource/hashme.txt');
  
  const orig = encodeForFileName(readFileSync(path), 'salty');
  const mute = encodeForFileName(Buffer.concat([
    readFileSync(path), 
    Buffer.from('hello')
  ]), 'salty');

  expect(orig).not.toBe(mute);
});
