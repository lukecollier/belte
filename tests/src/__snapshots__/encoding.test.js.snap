// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash is consistent for filesnames 1`] = `"qnwBADXwrvvXARNr"`;
